[Exclusivity] Diagnose when noescape closure is passed via reabstraction thunk

devincoughlin opened the pull request:
This fixes a serious false negative in static exclusivity enforcement when
a noescape closure performs an access that conflicts with an in-progress access
but is not reported because the closure is passed via a reabstraction thunk.

When diagnosing at a call site, the enforcement now looks through partial
applies that are passed as noescape arguments. If the partial apply takes
an argument that is itself a noescape partial apply, it recursively checks
the captures for that partial apply for conflicts and diagnoses if it finds one.

This means, for example, that the compiler will now diagnose when there is a
conflict involving a closure with a concrete return type that is passed to a
function that expects a closure returning a generic type. To preserve source
compatibility these diagnostics are emitted as warnings for now. The intent is
that they will be upgraded to errors in a future version of Swift.

rdar://problem/35215926, SR-6103

devincoughlin commented:
@swift-ci Please Test Source Compatibility

atrick commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -965,22 +965,33 @@ static PartialApplyInst *lookThroughForPartialApply(SILValue V) {
+  if (isa<FullApplySite>(AS) && Callee && !Callee->empty()) {
     // Check for violation with directly called closure
-    checkForViolationWithCall(Accesses, Callee, 0, FAS.getArguments(), ASA,
+    checkForViolationWithCall(Accesses, Callee, 0, AS.getArguments(), ASA,
```
Is '0' still a valid startIndex?

devincoughlin replied:
It is because that site is only used for full applies of directly called closures (it is not the recursive base case). I've reorganized the code to remove this logic from the recursive call to try to make this more clear.

atrick commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -965,22 +965,33 @@ static PartialApplyInst *lookThroughForPartialApply(SILValue V) {
+    // themselves no escape closures. This looks through reabstraction thunks.
+    checkForViolationsInNoEscapeClosures(Accesses, PAI, ASA,
+                                         ConflictingAccesses);
+
```
It seems to me that `checkForViolationWithCall()` will be called twice for each closure passed to a call. Once here when we recurse, and again below.

devincoughlin replied:
I've added an explicit check for whether the call is to a reabstraction thunk to prevent calling checkForViolationWithCall() on reabstraction thunks and instead perform the recursive case only.

devincoughlin commented:
I've updated this PR to try to make the recursion logic more clear and also to avoid directly diagnosing on the reabstraction thunks. Now, for reabstraction thunks we *only* call the recursive case, which diagnoses the closure passed to the reabstraction thunks.

I've also added a bunch more source-level tests for various kinds of thunks, added some SIL-level tests, and changed several existing SIL tests to use convert_function to apply the noescape function attribute.

@atrick Would you be willing to take another look?

devincoughlin committed:
[Exclusivity] Diagnose when noescape closure is passed via reabstraction thunk

This fixes a serious false negative in static exclusivity enforcement when
a noescape closure performs an access that conflicts with an in-progress access
but is not reported because the closure is passed via a reabstraction thunk.

When diagnosing at a call site, the enforcement now looks through partial
applies that are passed as noescape arguments. If the partial apply takes
an argument that is itself a noescape partial apply, it recursively checks
the captures for that partial apply for conflicts and diagnoses if it finds one.

This means, for example, that the compiler will now diagnose when there is a
conflict involving a closure with a concrete return type that is passed to a
function that expects a closure returning a generic type. To preserve source
compatibility these diagnostics are emitted as warnings for now. The intent is
that they will be upgraded to errors in a future version of Swift.

rdar://problem/35215926, SR-6103

devincoughlin commented:
Fixed a merge conflict involving the rename of getNumArguments().

devincoughlin commented:
@swift-ci Please smoke test and merge
